hail/src/main/resources/include/hail/SparkUtils.h

```diff
@@ -14,18 +14,126 @@ namespace hail {
+        SparkEnv * senv_;
+        jobject baos_;
+        jobject jctx_;
+        EltEncoder eltEnc_;
+        JNIEnv * env() { return senv_->env(); }
+
+      public:
+        ArrayEncoder(SparkEnv * senv, jobject baos) :
+        senv_(senv),
+        baos_(baos),
+        eltEnc_(std::make_shared<OutputStream>(*(senv->up()), baos)) { }
+
+        jobject encode(const char * ctxs) {
+          auto len = ArrayImpl::load_length(ctxs);
+          auto jctx = env()->NewObjectArray(len, senv_->byteArrayClass_, nullptr);
+          for (int i=0; i<len; ++i) {
+            if (ArrayImpl::is_element_missing(ctxs, i)) {
+              throw new FatalError("context cannot be missing");
+            } else {
+              env()->CallVoidMethod(baos_, senv_->OutputStream_reset_);
+              eltEnc_.encode_row(ArrayImpl::load_element(ctxs, i));
+              eltEnc_.flush();
+              env()->SetObjectArrayElement(jctx, i, env()->CallObjectMethod(baos_, senv_->OutputStream_toByteArray_));
+            }
+          }
+          env()->CallVoidMethod(baos_, senv_->OutputStream_reset_);
+          return jctx;
+        }
+    };
+
+    template <typename EltDecoder, typename ArrayBuilder>
+    class ArrayDecoder {
+      private:
+        SparkEnv * senv_;
+        JNIEnv * env() { return senv_->env(); }
+
+      public:
+        ArrayDecoder(SparkEnv * senv) :
+        senv_(senv) { }
+
+        const char * decode(Region * region, jobjectArray elements) {
+          auto len = env()->GetArrayLength(elements);
+          ArrayBuilder builder { len, region };
+          builder.clear_missing_bits();
+          for (int i=0; i<len; ++i) {
+            auto bais = env()->NewObject(senv_->bais_, senv_->InputStream_constructor_, env()->GetObjectArrayElement(elements, i));
+            EltDecoder elt_dec { std::make_shared<InputStream>(*(senv_->up()), bais) };
+            auto off = elt_dec.decode_row(region);
+            builder.set_element(i, off);
+            env()->DeleteLocalRef(bais);
+          }
+          return builder.offset();
+        }
+    };
+
+    template <typename CtxEncoder, typename GlobalEncoder, typename ResultDecoder>
+    const char * compute_distributed_array(RegionPtr region, const char * modID, const char * fname, const char * ctxs, const char * globals) {
+      auto baos = env()->NewObject(baos_, OutputStream_constructor_);
+      auto os = std::make_shared<OutputStream>(up_, baos);
+      GlobalEncoder globalEnc { os };
+      globalEnc.encode_row(globals);
+      globalEnc.flush();
+      auto jglobals = env()->CallObjectMethod(baos, OutputStream_toByteArray_);
+
+      CtxEncoder ctxEnc { this, baos };
+      auto jctxs = env()->NewLocalRef(ctxEnc.encode(ctxs));
+
+      jobject jmodID = env()->NewStringUTF(modID);
+      jobject jfname = env()->NewStringUTF(fname);
+      auto res = env()->NewGlobalRef(env()->CallObjectMethod(spark_utils_, parallelize_compute_collect_, jmodID, jfname, jctxs, jglobals));
```

Comment:
you're right--it shouldn't have to be. We shouldn't really need `GlobalRef`s in this part of the code, since we're generally going to be using these jobjects from the same c++ scope. I think I changed it when I was debugging at some point and forgot to change it back.